### DISCUSSION TITLE
Not sure how to validate in Frontend

### DISCUSSION THREAD

**Discussion by Seulasslintan** [mosparo/mosparo#123] [Q&A] [answered]
Hi,

I recently started using mosparo, and I'm having trouble understanding how the Frontend receives information about the validation status of the form. Since I'm working on my own HTML, JS, PHP project, I can't use a plugin.

I've figured out how the validation is processed on the server-side, but I'm unsure how to get the validation result on the Frontend. Currently, I have a button that should lead to the next page after successful validation. On that page, I've implemented the mosparo box for form validation, but I'm unsure how to retrieve the response from the plugin.

I've already added "onCheckForm" to the plugin's options during initialization, but I'm not sure how to proceed from here.

Any help would be greatly appreciated. Thanks and regards.

**Comment by zepich** [comment]
Hi @Seulasslintan 

Thank you very much for your question. Great to hear that you use mosparo!

The validation will happen when you click the mosparo checkbox. Then the data in the form will be collected and sent to your mosparo installation. mosparo will check the content based on the rules, and if everything is okay, the browser will get the result of the check. As a result, mosparo returns the validation token with which the server can later verify the form data. The mosparo JavaScript you included to display the box will put the validation token from mosparo into a hidden form field in your form. After the validation is executed, the callback `onCheckForm` is called to let you do other things. 

After the validation is done, you need to submit the form so the backend (server-side) can verify the data with mosparo (you could do that by submitting the form with the `onCheckForm` callback automatically or by waiting for the user to click the submit button). If you need the validation token in the `onCheckForm` callback, you could get it with JavaScript from the form element (but honestly, I don't think you need the token outside of your form).

The whole process looks like this:

1. The user fills out the form
2. The user clicks the mosparo checkbox
3. The mosparo JavaScript will send the form data to mosparo and validate the data (check the data for spam)
4. The user clicks the submit button (or your code can do that automatically after validation in the `onCheckForm` callback)
5. Your backend (for example, PHP) must verify the submission by verifying that nobody manipulated the form data after the validation)
6. If mosparo verified the data, everything is good, and you can do with the data what you want (send by email, store in a database)

In the submitted form data to your backend (for example, PHP), you will receive two additional keys in the request: _mosparo_submitToken and _mosparo_validationToken. You need These two data to verify the form data by using the PHP API client, for example (or by calling the verification API directly).

Please let me know if that helps you to understand the process, or if I misunderstood you, or if you have any other questions.

Kind regards,

zepich

**Reply by Seulasslintan** [reply]
Hello zepich,

Thank you very much for your prompt reply.

I have understood the process itself, including the inclusion of the button and the confirmation functionality. However, I'm wondering how to retrieve information from the checkbox or the mosparo object on the frontend side to determine if the validation was successful or not.

On my website, I check if all fields are filled out and I also need to know whether the validation result was successful before allowing the visitor to proceed to the next page.

I hope my explanation is clearer this time. Thank you for your assistance.

Best regards,
Seulasslintan

**Reply by zepich** [reply, marked as answer]
Hi @Seulasslintan 

Thank you very much for your response, and sorry for the misunderstanding.

Ah, now I understand. So, the callback `onCheckForm` gets the result as a boolean value. If the validation is successful (everything is good to submit), the parameter will be `true`, otherwise `false` (see here: https://github.com/mosparo/mosparo/blob/master/assets/mosparo-frontend.js#L336).

With that, you should be able to enable the button so the user can click it.

Additionally, if the validation was successful, the main mosparo HTML element (the one for which you specify the ID in the initialization) gets the class `mosparo__checked`. You can check if this class is added to the element, and if yes, you know that the validation was successful.

You can also check if the checkbox (`input[type="checkbox"]`) inside the mosparo box is `checked`. The checkbox is only checked if the validation was successful. Otherwise, it will be unchecked and block the submission of the form (because the field is set as required, you cannot submit the form until the checkbox is checked).

Does this help with your integration?

Kind regards,

zepich

**Reply by Seulasslintan** [reply]
Hi @zepich,

Thank you once again for your swift response. Your answer has been very helpful in understanding the process and integrating mosparo into my project.

Wishing you and the team at mosparo a great day!

Best regards,
Seulasslintan

**Reply by zepich** [reply]
Thank you very much! 

I'm happy to help!

Kind regards,

zepich
